execute/plugin_functions.go

```diff
@@ -208,6 +202,117 @@ func filterOutExecutedMessages(
 	return filtered, nil
 }
 
+// truncateObservation truncates the observation to fit within the given maxSize after encoding.
+// It removes data from the observation in the following order:
+// For each chain, remove commit reports one by one, if the encoded observation is still too large,
+// remove the entire chain.
+// Keep repeating this process until the encoded observation fits within the maxSize or there's only
+// one chain with one report left
+// Note: This function doesn't split one report into multiple parts.
+func truncateObservation(
+	observation *exectypes.Observation,
+	maxSize int,
+) {
+	encodedObs, err := observation.Encode()
+	if err != nil {
+		return
+	}
+
+	chains := maps.Keys(observation.CommitReports)
+	sort.Slice(chains, func(i, j int) bool {
+		return chains[i] < chains[j]
+	})
+
+	// If the encoded observation is too large, start filtering data.
+	for len(encodedObs) > maxSize {
+		for _, chain := range chains {
+			if len(observation.CommitReports[chain]) > 1 {
+				truncateLastCommit(chain, observation)
+			} else {
+				truncateChain(chain, observation)
+			}
+			chains = maps.Keys(observation.CommitReports)
+			break
+		}
+		// If there's only one chain with one report left, stop.
+		if len(observation.CommitReports) == 1 && len(observation.CommitReports[chains[0]]) == 1 {
+			break
+		}
+		encodedObs, err = observation.Encode()
+		if err != nil {
+			return
+		}
+	}
+}
+
+// truncateLastCommit removes the last commit from the observation.
+// errors if there are no commits to truncate.
+func truncateLastCommit(
+	chain cciptypes.ChainSelector,
+	observation *exectypes.Observation,
+) bool {
+	commits := observation.CommitReports[chain]
+	if len(commits) == 0 {
+		return false
+	}
+	lastCommit := commits[len(commits)-1]
+	// Remove the last commit from the list.
+	commits = commits[:len(commits)-1]
+	observation.CommitReports[chain] = commits
+	for seqNum, msg := range observation.Messages[chain] {
+		if lastCommit.SequenceNumberRange.Contains(seqNum) {
+			// Remove the message from the observation.
+			delete(observation.Messages[chain], seqNum)
+			// Remove the token data from the observation.
+			delete(observation.TokenData[chain], seqNum)
+			// Remove costly messages
+			for i, costlyMessage := range observation.CostlyMessages {
+				if costlyMessage == msg.Header.MessageID {
+					observation.CostlyMessages = append(observation.CostlyMessages[:i], observation.CostlyMessages[i+1:]...)
+				}
+			}
+			// Leaving Nonces untouched
```

Comment:
I think this would be confusing for debugging, why not remove it?

Reply:
Size is negligible compared to the rest of the objects and removing is complex to do. It shouldn't be really confusing IMO.